### ISSUE TITLE
How to proceed with update that will cause FediFetcher to be very slow for a few executions?

### ISSUE THREAD

**Issue by nanos** [nanos/FediFetcher#32]
I'm currently working on a new feature for FediFetcher which will backfill the profiles of any users mentioned in posts on your home timeline. If your home timeline contains posts by accounts you don't follow (e.g. because you are following hashtags), these accounts will also be backfilled.

Due to the large number of posts involved, this can take a lot of time for the first execution. Because FediFetcher caches profiles that it has backfilled, it'll then catch up relatively quickly in subsequent executions (as long as you don't leave too long gaps between executions), and once it had caught up, I noticed no significant speed difference.

In my testing, when fetching context for the last 400 posts in my home timeline, these were the execution times: 1st run after the update: 4.5 hours. 2nd run: 1.5 hours. 3rd run: 45 min. 4th and subsequent runs: < 5 min.

Obviously your mileage will vary, depending particularly on how many accounts you don't follow appear in your timeline.

Now my question: How should I deal with this update, particularly for users of GitHub Actions, who'll just get the update without any action on their side?

I could either make this new behaviour optional, and disable it on the first run, or I could just enable it for everyone. But for users of GitHub Actions, that would mean that the first run after release would suddenly take several hours, without obvious reasons on why this happened. 

My preferred option is to just ship it, and let every admin deal with it. I don't really want to make this optional, let alone turn it off by default, because I think it's super helpful.

But I'm also conscious that I don't want to alienate admins who might trust me with this script.

So, particularly if you are using FediFetcher as a GitHub Action, I'd love to hear from you: Should this be enabled by default, or not?

**Comment by nanos** [comment]
@quicoto tagging you here, because you observed this because I made a mistake earlier 😬 I'd appreciate your feedback!

**Comment by quicoto** [comment]
It's a tough one because there's no way to notify the GitHub Action users that will will happen.

I noticed because the scheduled 10 min runs started to fail because an ongoing task.

I get that it's useful and I do want that feature. Maybe we need to throttle it a bit as a default for a few days / weeks.

What if we ship it but it only backfills a small set of users, say 5 users at a time (or last 10 posts, no idea). Then, let's say, after 24 hours you could bump the number. So it would be progressive. The GitHub Actions users wouldn't have to do anything as they would fetch the release from the main repo. Could this be a compromise?

I do not know what's the intended configuration you had in mind but 4.5 hours feels like a lot for a single run.

**Comment by gashead76** [comment]
I do use GitHub actions and personally I would be ok with the update having a super long execution time for the first couple of runs after this update because the way I use Mastodon I have lots of non-followed folks in my timeline from following hashtags. I would simply like to be notified (which now I have been). However, I think there might be at least a few admins out there who would not like that behavior.

How about instead of fetching the last 400 (or even much less), perhaps have the new feature start from this new update point forward? Would that be a feasible compromise?

In any case, I appreciate the work you've put into this and I look forward to fetching even more content with this new feature!

**Comment by edoswald** [comment]
I'm gonna have to go with quicoto here on the gradual release. Could be other benefits, too, if something goes wrong. I would consider a run "too long" if it would bleed in the next and other scheduled runs. I know most people run this from Github, but this would be something that might deserve a nag right before those that run via command line set it up, acknowledging they understand that the first few runs may be unusually long (might save you from some "my script is running forever" support questions).

I wanted to know why the Github script occasionally failed, but that's been answered above. Not an issue to me as I'd rather save the computing power to maintain QoS with the expected run for Mastodon post April 1st. 

Either way, fantastic work. Has made my little instance feel like mastodon.social. hopefully the main devs see why this is important sooner rather than later, so it's built right in.

**Comment by p37307** [comment]
What the hey. I'll give it a shot and get back to you from Windows command line).

**Comment by nanos** [comment]
Do remember to download your artefacts and pause your workflow though.

**Comment by p37307** [comment]
First thing I did was disabled my task schedule. Let's see how it goes. Curious how long it will take because I host the #HashtagGames at https://oldfriends.live/@hashtaggames and any given night I interact with many players and hundreds game plays over dozens of instances. I'm not sure what that means volume wise when it backfills.

**Comment by nanos** [comment]
That's a really interesting one @p37307. 

Have you tried this in the `dev` branch by any chance? I'd love to know how you get on during these periods of peak activity.

I'm running my cron job every 2 minutes at the moment, and usually each execution takes between 2-3 minutes, so it's really not bad at all for myself. But I wonder how it would behave during such short boosts of activity.

**Comment by p37307** [comment]
@nanos I am running the dev branch. First runs. So far, the first run was 3.5 hours. 2nd run was only two hours. 3rd run, which occurred during and after the 9 oclock hashtag game, went up to 2 hours and 45 minutes. I am on the 4th run now. I wished there was a logging option. I do one new hashtag game per day at 9pm eastern time. So far, tonight's hashtag game interacted with around 70 players, around 400 game plays, not counting replies, etc., not counting ongoing games. They seem to go on for days after the initial play. Between 2nd run and third run, 45 minutes was added to the scan.

**Comment by nanos** [comment]
@p37307  in terms of logging: FediFetcher logs so much, that I’m actually tempted to remove some of it. 

It just logs to stdout, but you could just redirect that output to a file.

Another option is to use the `—on-start` and `—on-done` parameters. I use these with healthchecks.io and that gives me a good glanceable look at how long FediFetcher takes for each execution:

<img width="955" alt="image" src="https://user-images.githubusercontent.com/5182595/229276936-e9df86f2-d2de-4714-ae2b-3a9d3e7b4793.png">


overall my cronjob looks like this, which creates a log file per execution (and yes, I do have a second cronjob that clears out the logs after a while), and logs execution times with healthchecks.io for easy monitoring:

```
*/2 * * * * python3 find_posts.py --access-token= --server=mstdn.thms.uk --on-done="https://hc-ping.com/***" --on-start="https://hc-ping.com/***/start" > /home/Michael/fedifetcher-logs/output-`/bin/date +\%Y\%m\%d\%H\%M\%S`.log 2>&1
```

**Comment by nanos** [comment]
Thank you all, for your input!

I have just released v4.1.7, which includes these changes.

To keep the execution time of FediFetcher under control, FediFetcher will backfill at most 30 users per execution. In my testing, I was able to backfill about 2 users per minute on average, meaning that this part of the script should run for no more than 15 minutes, which seems acceptable to me.

If you wanted to turn this new feature off, you can set the `BACKFILL_MENTIONED_USERS` / `--backfill-mentioned-users` option to `0`.
